### PR TITLE
resolves #456 add a root-folder in distribution bundles (zip files)

### LINKED ISSUE #456: Add a root-folder in distribution bundles (zip files)
Unzipping a precompiled binaries messes up the current folder with all kind of (installation) clutter. Better to have a root-folder like `asciidoctor-web-pdf-x.y.z` that contains everyhing.

### PR DESCRIPTION
**Pull request by wh81752**
Changes for issue #456 . This PR also contains an alternative solution for #459 (terminating Node.js script).

In order to test my changes to GitHub Actions I create this PR. I guess I should test upfront in my own fork. Is that the way to do it?

### CODE REVIEW

**Comment by wh81752** [comment]
Ok, after some fiddling with GitHub Actions -- especially how to extract version and share version between workflows -- it starts to looks good. The package contents are now within a versioned root folder, i.e.

```
$ export version=1.0.0-alpha.13         # whatever the next released version will be
$ wget https://github.com/Mogztter/asciidoctor-web-pdf/releases/download/v1.0.0-alpha.12/asciidoctor-web-pdf-linux-v${version}.zip
$ unzip asciidoctor-web-pdf-linux-v${version}.zip
$ ls asciidoctor-web-pdf-linux-${version}
assets/		chromium/	css/		examples/	fonts/     asciidoctor-web-pdf*
```

There is one remaining problem:
1. The version in the published package name starts with a "v" while the version of the root folder is without that "v"

File: tasks/prepare-binaries.js [hunk 1]
```diff
@@ -2,6 +2,8 @@ const path = require('path')
 const fs = require('fs')
 const fsExtra = require('fs-extra')
 const readline = require('readline')
+
+process.stdin.unref()
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:6]:
```suggestion
// When creating a readline.Interface using stdin as input, the program will not terminate until it receives EOF (Ctrl+D on Linux/macOS, Ctrl+Z followed by Return on Windows).
// If you want your application to exit without waiting for user input, you can unref the standard input stream.
// Reference: https://nodejs.org/api/readline.html#readline_readline
process.stdin.unref()
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -22,8 +22,17 @@ jobs:
...
+      version: ${{ steps.output_version.outputs.version }}
     steps:
     - uses: actions/checkout@v2
+    - name: Extract version
+      id: extract_version
+      uses: Saionaro/extract-package-version@v1.1.1
+    - name: Print version
+      run: echo ${{ steps.extract_version.outputs.version }}
+    - id: output_version
+      run: echo "::set-output name=version::${{ steps.extract_version.outputs.version }}"
```
Comment (Mogztter) [inline .github/workflows/build.yml:35]: For such a simple task, we can use:


```suggestion
    - id: package_info
      run: |
        VERSION=$(node -e 'console.log(require("./package.json").version)')
        echo "::set-output name=version::$VERSION"
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -22,8 +22,17 @@ jobs:
             node-version: '14.x'
             primary: true
     runs-on: ${{ matrix.os }}
+    outputs:
+      version: ${{ steps.output_version.outputs.version }}
```
Comment (Mogztter) [inline .github/workflows/build.yml:26]: Neat trick 👌🏻 

```suggestion
      version: ${{ steps.package_info.outputs.version }}
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -60,7 +69,7 @@ jobs:
     - name: Smoke test
       run: |
         unzip asciidoctor-web-pdf-linux.zip -d out
-        ./out/asciidoctor-web-pdf ./out/examples/document/basic-example.adoc -o basic.pdf
+        ./out/asciidoctor-web-pdf-linux-${{ steps.extract_version.outputs.version }}/asciidoctor-web-pdf ./out/asciidoctor-web-pdf-linux-${{ steps.extract_version.outputs.version }}/examples/document/basic-example.adoc -o basic.pdf
```
Comment (Mogztter) [inline .github/workflows/build.yml:72]: We can use `${{ steps.package_info.outputs.version }}` no?


```suggestion
        ./out/asciidoctor-web-pdf-linux-${{ steps.package_info.outputs.version }}/asciidoctor-web-pdf ./out/asciidoctor-web-pdf-linux-${{ steps.package_info.outputs.version }}/examples/document/basic-example.adoc -o basic.pdf
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -86,6 +95,6 @@ jobs:
       - name: Smoke test
         run: |
           unzip asciidoctor-web-pdf-win.zip -d out
-          out\asciidoctor-web-pdf.exe out\examples\document\basic-example.adoc -o basic.pdf
+          out\asciidoctor-web-pdf-win-${{needs.build.outputs.version}}\asciidoctor-web-pdf.exe out\asciidoctor-web-pdf-win-${{needs.build.outputs.version}}\examples\document\basic-example.adoc -o basic.pdf
```
Comment (Mogztter) [inline .github/workflows/build.yml:98]:
```suggestion
          out\asciidoctor-web-pdf-win-${{ needs.build.outputs.version }}\asciidoctor-web-pdf.exe out\asciidoctor-web-pdf-win-${{ needs.build.outputs.version }}\examples\document\basic-example.adoc -o basic.pdf
```

File: tasks/prepare-binaries.js [hunk 6]
```diff
@@ -75,9 +79,10 @@ async function archive (platforms) {
     const archive = archiver('zip', {
       zlib: { level: 9 } // Maximize compression
     })
-
+    const archiveName = `${appName}-${platform}`
+    const rootFolder = `${appName}-${platform}-${version}`
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:83]:
```suggestion
    const rootFolder = `${archiveName}-${version}`
```

File: tasks/prepare-binaries.js [hunk 7]
```diff
@@ -86,8 +91,8 @@ async function archive (platforms) {
     })
     // pipe archive to file stream
     archive.pipe(zipOut)
-    // recursively add directory to _root_ of zip
-    archive.directory(path.join(buildDirPath, platform), false)
+    // recursively add directory to folder ${appName}-${platform}-${version}
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:94]:
```suggestion
    // recursively add directory to folder ${archiveName}-${version}
```

File: tasks/prepare-binaries.js [hunk 8]
```diff
@@ -24,6 +26,8 @@ const platforms = {
   win: { target: 'node12-win-x64', suffix: '.exe', puppeteerPlatform: 'win64' }
 }
 
+const version = require('../package.json').version
+
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:30]:
```suggestion
const version = require('../package.json').version
```

File: tasks/prepare-binaries.js (hunk 1 above)
Comment (Mogztter) [reply tasks/prepare-binaries.js:6]: Can we close it after `readline.createInterface`?

**Comment by Mogztter** [comment]
A few nitpicks but otherwise it looks good 👍🏻 
Thanks for taking the time to fiddle with GitHub Actions.

> The version in the published package name starts with a "v" while the version of the root folder is without that "v"

I don't have a strong opinion whether we should keep the "v" prefix or remove it but it's important that we stay consistent.

File: tasks/prepare-binaries.js (hunk 1 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:6]: Yes of course.

**Comment by wh81752** [comment]
Btw, would it not be more consistent to have a smoke test for linux similiar to smoke tests for windows and macos?

**Comment by Mogztter** [comment]
> Btw, would it not be more consistent to have a smoke test for linux similiar to smoke tests for windows and macos?

Why not, the goal was to save some time but it's negligible.

**Comment by wh81752** [comment]
How about merging this PR, or is there a problem left?

**Comment by Mogztter** [comment]
No it looks good, I will review one more time, potentially make small changes and squash commits before merging 😉

**Comment by Mogztter** [comment]
And merged, thanks @wh81752 🤗

**Pull request merged**